cmd/thanos/receive.go

```diff
@@ -613,22 +615,35 @@ func startTSDBAndUpload(g *run.Group,
 				if !ok {
 					return nil
 				}
-				dbUpdatesStarted.Inc()
-				level.Info(logger).Log("msg", "updating storage")
 
-				if err := dbs.Flush(); err != nil {
-					return errors.Wrap(err, "flushing storage")
-				}
-				if err := dbs.Open(); err != nil {
-					return errors.Wrap(err, "opening storage")
-				}
-				if upload {
-					uploadC <- struct{}{}
-					<-uploadDone
+				// When using Ketama as the hashring algorithm, there is no need to flush the TSDB head.
+				// If new receivers were added to the hashring, existing receivers will not need to
+				// ingest additional series.
+				// If receivers are removed from the hashring, existing receivers will only need
+				// to ingest a subset of the series that were assigned to the removed receivers.
+				// As a result, changing the hashring produces no churn, hence no need to force
+				// head compaction and upload.
+				flushHead := !initialized || hashringAlgorithm != receive.AlgorithmKetama
+				if flushHead {
```

Comment:
One thing we should be careful about now is that removing a receiver from the hashring will not flush the head, which means head compaction will happen only when we GC the TSDB after the retention period. With a retention of 15d, this means samples will remain in the head for the entire 15d period.

We should maybe run manual compactions every 2 hours as part of the GC goroutine?

Reply:
Hm, is this right? Are we talking about head compaction? Does retention period affect head compaction? I was under the impression head compaction still occurs in ~2 hour intervals.

Reply:
Yes, head compaction occurs every 2 hours, but it is triggered by:
* a sample being appended, or
* prometheus compactor running in background.

So if we remove a tenant from a receiver, none of those will happen.

Reply:
One concern I have with compacting the head every N hours is that I am not sure whether it can create out of order samples in certain cases.

For example, a tenant has an outage and fails to send data for 4 hours. These samples can be kept in the prometheus agent with proper configuration: https://github.com/prometheus/prometheus/blob/a1fcfe62dbe82c6292214f50ee91337566b0d61b/tsdb/agent/db.go#L87-L88. 

Receiver decides to compact the head and flush it to disk because it hasn't seen any new samples for a while. So now when the tenant comes back online and sends all samples from the backlog, some of those samples can be older than the new head min-time and I assume they will be dropped as OOO.

Reply:
That is a great point. We already have this situation with separate Router + Ingester.

Repro would be:

* X receivers, hashring that accidentally shard series in a way that tenant A and tenant B is on receiver e.g. 10th.
* Resharding happens. With this PR  we don't flush. Let's say sharding changes to receiver 10th does not receiver tenant B series anymore.
* The tenant B will stay in memory forever as no sample appends

To me this situation is fine. We never grow memory uncontrollably, plus we have this tenant removal when unused logic you added @fpetkovski 

I would not add anything else here. Am I missing anything @matej-g @fpetkovski ?

Reply:
I think tenant GC will cover us for now. Users can use the retention period to control when they want GC to happen.